extensions/auth/opa/impl/src/main/java/org/apache/polaris/extension/auth/opa/token/FileBearerTokenProvider.java

```diff
@@ -0,0 +1,307 @@
+    } finally {
+      lock.writeLock().unlock();
+    }
+  }
+
+  private boolean shouldRefresh() {
+    return clock.instant().isAfter(nextRefresh);
+  }
+
+  private void refreshToken() {
+    lock.writeLock().lock();
+    try {
+      // Double-check pattern - another thread might have refreshed while we waited for the lock
+      if (!shouldRefresh()) {
+        return;
+      }
+
+      String newToken = loadTokenFromFile();
+
+      // If we couldn't load a token and have no cached token, this is a fatal error
+      if (newToken == null && cachedToken == null) {
+        throw new RuntimeException(
+            "Unable to load bearer token from file: "
+                + tokenFilePath
+                + ". This is required for OPA authorization.");
+      }
+
+      // Only update cached token if we successfully loaded a new one
+      if (newToken != null) {
+        cachedToken = newToken;
+      }
+      // If newToken is null but cachedToken exists, we keep using the cached token
+
+      lastRefresh = clock.instant();
+
+      // Calculate next refresh time based on current token (may be cached)
+      nextRefresh = calculateNextRefresh(cachedToken);
+
+      logger.debug(
+          "Token refreshed from file: {} (token present: {}), next refresh: {}",
+          tokenFilePath,
+          cachedToken != null && !cachedToken.isEmpty(),
+          nextRefresh);
+
+    } finally {
+      lock.writeLock().unlock();
+    }
+  }
+
+  /** Calculate when the next refresh should occur based on JWT expiration or fixed interval. */
+  private Instant calculateNextRefresh(@Nullable String token) {
+    if (token == null || !jwtExpirationRefresh) {
+      // Use fixed interval
+      return lastRefresh.plus(refreshInterval);
+    }
+
+    // Attempt to parse as JWT and extract expiration
+    Optional<Instant> expiration = getJwtExpirationTime(token);
+
+    if (expiration.isPresent()) {
+      // Refresh before expiration minus buffer
+      Instant refreshTime = expiration.get().minus(jwtExpirationBuffer);
+
+      // Ensure refresh time is in the future and not too soon (at least 1 second)
+      Instant minRefreshTime = clock.instant().plus(Duration.ofSeconds(1));
+      if (refreshTime.isBefore(minRefreshTime)) {
+        logger.warn(
+            "JWT expires too soon ({}), using minimum refresh interval instead", expiration.get());
+        return lastRefresh.plus(refreshInterval);
+      }
+
+      logger.debug(
+          "Using JWT expiration-based refresh: token expires at {}, refreshing at {}",
+          expiration.get(),
+          refreshTime);
+      return refreshTime;
+    }
+
+    // Fall back to fixed interval (token is not a valid JWT or has no expiration)
+    logger.debug("Token is not a valid JWT or has no expiration, using fixed refresh interval");
+    return lastRefresh.plus(refreshInterval);
+  }
+
+  @Nullable
+  private String loadTokenFromFile() {
+    int attempts = 0;
+    long deadlineMs =
```

Comment:
Hm, this would block all requests.
What I had in mind is to try only once (as before) and yield the previous token in case of an error.